File: im-desktop/src/services/at-mention-service.ts
import axios from 'axios';
import { AtMention, AtMentionRequest, AtMentionSettings, AtMentionListResponse, AtMentionCountResponse } from '../types/at-mention';
import { useAuthStore } from '../stores/auth-store';

const API_BASE = '/api/at-mention';

function getHeaders() {
  const token = useAuthStore.getState().token;
  return { Authorization: `Bearer ${token}` };
}

/**
 * 处理消息@提及
 */
export async function processMentions(request: AtMentionRequest): Promise<AtMention[]> {
  const response = await axios.post(`${API_BASE}/process`, request, { headers: getHeaders() });
  return response.data.data;
}

/**
 * 获取用户@提及列表
 */
export async function getMentionList(userId: number, page = 0, size = 20): Promise<{ data: AtMention[], totalPages: number, totalElements: number }> {
  const response = await axios.get<AtMentionListResponse>(`${API_BASE}/list`, {
    params: { userId, page, size },
    headers: getHeaders()
  });
  return {
    data: response.data.data,
    totalPages: response.data.totalPages,
    totalElements: response.data.totalElements
  };
}

/**
 * 获取未读@提及数量
 */
export async function getUnreadCount(userId: number): Promise<number> {
  const response = await axios.get<AtMentionCountResponse>(`${API_BASE}/unread-count`, {
    params: { userId },
    headers: getHeaders()
  });
  return response.data.data;
}

/**
 * 获取群聊未读@提及数量
 */
export async function getUnreadCountInRoom(userId: number, roomId: number): Promise<number> {
  const response = await axios.get<AtMentionCountResponse>(`${API_BASE}/unread-count/room`, {
    params: { userId, roomId },
    headers: getHeaders()
  });
  return response.data.data;
}

/**
 * 标记@提及已读
 */
export async function markAsRead(userId: number, mentionIds: number[]): Promise<number> {
  const response = await axios.post(`${API_BASE}/mark-read`, mentionIds, {
    params: { userId },
    headers: getHeaders()
  });
  return response.data.updated;
}

/**
 * 标记群聊内所有@已读
 */
export async function markAllAsReadInRoom(userId: number, roomId: number): Promise<number> {
  const response = await axios.post(`${API_BASE}/mark-read/room`, {}, {
    params: { userId, roomId },
    headers: getHeaders()
  });
  return response.data.updated;
}

/**
 * 获取@提及设置
 */
export async function getMentionSettings(userId: number): Promise<AtMentionSettings> {
  const response = await axios.get(`${API_BASE}/settings`, {
    params: { userId },
    headers: getHeaders()
  });
  return response.data.data;
}

/**
 * 更新@提及设置
 */
export async function updateMentionSettings(userId: number, settings: AtMentionSettings): Promise<AtMentionSettings> {
  const response = await axios.put(`${API_BASE}/settings`, settings, {
    params: { userId },
    headers: getHeaders()
  });
  return response.data.data;
}

File: im-mobile/lib/services/at_mention_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/at_mention.dart';

class AtMentionService {
  static const String _baseUrl = '/api/at-mention';
  final String _token;

  AtMentionService(this._token);

  Map<String, String> get _headers => {
    'Authorization': 'Bearer $_token',
    'Content-Type': 'application/json',
  };

  Future<List<AtMention>> processMentions({
    required int messageId,
    required int senderUserId,
    List<int>? mentionedUserIds,
    bool isAtAll = false,
    String? conversationId,
    int? roomId,
    String? messagePreview,
  }) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/process'),
      headers: _headers,
      body: jsonEncode({
        'messageId': messageId,
        'senderUserId': senderUserId,
        'mentionedUserIds': mentionedUserIds,
        'isAtAll': isAtAll,
        'conversationId': conversationId,
        'roomId': roomId,
        'messagePreview': messagePreview,
      }),
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return (data['data'] as List).map((e) => AtMention.fromJson(e)).toList();
  }

  Future<List<AtMention>> getMentionList(int userId, {int page = 0, int size = 20}) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/list').replace(queryParameters: {
        'userId': userId.toString(),
        'page': page.toString(),
        'size': size.toString(),
      }),
      headers: _headers,
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return (data['data'] as List).map((e) => AtMention.fromJson(e)).toList();
  }

  Future<int> getUnreadCount(int userId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/unread-count').replace(queryParameters: {
        'userId': userId.toString(),
      }),
      headers: _headers,
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return data['data'] ?? 0;
  }

  Future<int> getUnreadCountInRoom(int userId, int roomId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/unread-count/room').replace(queryParameters: {
        'userId': userId.toString(),
        'roomId': roomId.toString(),
      }),
      headers: _headers,
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return data['data'] ?? 0;
  }

  Future<int> markAsRead(int userId, List<int> mentionIds) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/mark-read').replace(queryParameters: {
        'userId': userId.toString(),
      }),
      headers: _headers,
      body: jsonEncode(mentionIds),
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return data['updated'] ?? 0;
  }

  Future<int> markAllAsReadInRoom(int userId, int roomId) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/mark-read/room').replace(queryParameters: {
        'userId': userId.toString(),
        'roomId': roomId.toString(),
      }),
      headers: _headers,
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return data['updated'] ?? 0;
  }

  Future<AtMentionSettings> getMentionSettings(int userId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/settings').replace(queryParameters: {
        'userId': userId.toString(),
      }),
      headers: _headers,
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return AtMentionSettings.fromJson(data['data']);
  }

  Future<AtMentionSettings> updateMentionSettings(int userId, AtMentionSettings settings) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/settings').replace(queryParameters: {
        'userId': userId.toString(),
      }),
      headers: _headers,
      body: jsonEncode(settings.toJson()),
    );
    final data = jsonDecode(utf8.decode(response.bodyBytes));
    return AtMentionSettings.fromJson(data['data']);
  }
}
